pyuvdata/tests/__init__.py

```diff
@@ -5,19 +5,245 @@
+import warnings
 
 from astropy.utils import iers
 
 import pyuvdata.utils as uvutils
 
 __all__ = [
+    "check_warnings",
     "clearWarnings",
     "checkWarnings",
 ]
 
 
+class WarningsChecker(warnings.catch_warnings):
+    """
+    A context manager to check raised warnings.
+
+    Adapted from pytest WarningsRecorder and WarningsChecker.
+    """
+
+    def __init__(self, expected_warning, match):
+        super().__init__(record=True)
+        self._entered = False
+        self._list = []
+
+        assert isinstance(expected_warning, list)
+
+        msg = "exceptions must be derived from Warning, not %s"
+        if expected_warning is None:
+            expected_warning_list = expected_warning
+        elif isinstance(expected_warning, list):
+            for exc in expected_warning:
+                if not issubclass(exc, Warning):
+                    raise TypeError(msg % type(exc))
+            expected_warning_list = expected_warning
+        elif issubclass(expected_warning, Warning):
+            expected_warning_list = [expected_warning]
+        else:
+            raise TypeError(msg % type(expected_warning))
+
+        msg = "match must be a str, not %s"
+        if match is None:
+            match_list = None
+        elif isinstance(match, list):
+            for exc in match:
+                if not isinstance(exc, str):
+                    raise TypeError(msg % type(exc))
+            match_list = match
+        elif isinstance(match, str):
+            match_list = [match]
+        else:
+            raise TypeError(msg % type(match))
+
+        self.expected_warning = expected_warning_list
+        self.match = match_list
+
+    @property
+    def warnlist(self):
+        """The list of recorded warnings."""
+        return self._list
+
+    def __getitem__(self, i: int):
+        """Get a recorded warning by index."""
+        return self._list[i]
+
+    def __iter__(self):
+        """Iterate through the recorded warnings."""
+        return iter(self._list)
+
+    def __len__(self):
+        """The number of recorded warnings."""
+        return len(self._list)
+
+    def pop(self, cls):
+        """Pop the first recorded warning, raise exception if not exists."""
+        for i, w in enumerate(self._list):
+            if issubclass(w.category, cls):
+                return self._list.pop(i)
+        __tracebackhide__ = True
+        raise AssertionError("%r not found in warning list" % cls)
+
+    def clear(self) -> None:
+        """Clear the list of recorded warnings."""
+        self._list[:] = []
+
+    def __enter__(self):
+        if self._entered:
+            __tracebackhide__ = True
+            raise RuntimeError("Cannot enter %r twice" % self)
+        _list = super().__enter__()
+        # record=True means it's None.
+        assert _list is not None
+        self._list = _list
+        warnings.simplefilter("always")
+        warnings.filterwarnings("ignore", message="numpy.dtype size changed")
```

Comment:
should we add a small comment here about why we filter these for our tests?

Reply:
sure, will do.